File: 2_stack_queue_heap/이준우/택배 상자.kt
class Solution {
    fun solution(order: IntArray): Int {
        var answer = 0
        var n = 1 // 주 컨테어너에서 나온 박스의 현재 순번
        val stack = ArrayDeque<Int>() // 보조 컨테이너

        // 기사님이 원하는 박스의 순번
        for (want in order) {
            // 보조 컨테이너에 기사님이 현재 원하는 순번의 상자까지 담는다.
            while (stack.isEmpty() || stack.last() != want) {
                if (n > order.size) break

                stack.add(n)
                n++
            }

            // 보조 컨테이너의 현재 순번이 기사님이 원하는 순번이면 꺼낸다.
            if (stack.isNotEmpty() && stack.last() == want) {
                stack.removeLast()
                answer += 1
            } else { // 그렇지 않으면 해당 시뮬레이션을 종료한다. 
                break
            }
        }

        return answer
    }
}

File: 2_stack_queue_heap/이준우/택배 상자.py
def solution(order):
    answer = 0
    stack = []
    n = 1

    for want in order:
        while (not stack or want != stack[-1]) and n <= len(order):
            stack.append(n)
            n += 1

        if stack and stack[-1] == want:
            answer += 1
            stack.pop()
        else:
            break

    return answer


# 옛날 풀이
def solution(order):
    answer = 0
    order_index = 0
    stack = []

    for box in range(1, len(order) + 1):
        # 현재 박스를 스택에 넣음
        stack.append(box)

        # 스택의 탑이 필요한 박스와 일치하는 동안 계속 처리
        while stack and stack[-1] == order[order_index]:
            stack.pop()
            order_index += 1
            answer += 1

    return answer
